Add Consumer group data, with offsets and lag

cvcal opened the pull request:
This addresses issue https://github.com/yahoo/kafka-manager/issues/80

cvcal commented:
Two solutions to this problem that arrived on the same day (the walmart one, https://github.com/yahoo/kafka-manager/pull/109, and this one). 

Walmart doesn't use curator tree caches unlike the topics cache and also like my solution's consumer cache. This greatly reduces the amount of memory being used. It also adds automatic filtering of consumers that Kafka doesn't consider to be active. However, the consumer and offset code is minimally integrated into the existing code. Some features, like modifications to the topic view to add offsets and a list of consumers reading from the topic, are less feasible with this lack of integration.  

I am thinking of replacing the method I get consumer data with Walmart's code, to take advantage of the memory efficiency, while maintaining the integration into the actor system of the current kafka-manager. This will allow for a more complete integration of the consumer and offset information into the UI.

Please reply to let me know if this sounds like a good solution.

jisookim0513 commented:
I might be wrong, but I prefer your original approach of using consumer cache despite of the fact that it takes more memory than fetching data directly from Zookeeper. One of the things I like Kafka Manager is that it maintains a cache about Kafka state so it doesn't have to fetch data from Zookeeper every time I want to get some information about brokers or topics. I think the use of consumer cache fits better to the design of Kafka Manager, and could be better for future usage if I want to add a REST API to Kafka Manager to retrieve consumer information, in which case Kafka Manager wouldn't have to fetch information from Zookeeper but return value directly from its cache to the client. It's just my two cents.

cvcal commented:
I agree, that was why I used the cache in the first place, but I thought it was worth trying the other way to check if the performance didn't improve significantly. At first glance, it doesn't look like calling zookeeper directly makes a huge difference in the memory usage anyway. 

The problem might be more linked to having both the consumer descriptions and identities, which I did to follow the model of how the topic data was passed, but makes less sense for the consumers than for topics, as the data comes in from the treeCache much as it ends up being used. There isn't any JSON to read from, for one.

cvcal commented:
If I were to keep the filtering of inactive consumers (consumers that don't list any topics that it reads from) should I make it an option - say part of ClusterConfig?

patelh commented:
Can you add tests?

cvcal commented:
That definitely passed the tests on my machine, odd. 
I can add tests that simply respond that there aren't any consumers, but at least it would test that the call went through, etc. I'm not sure how to modify your SeededBroker to add a seeded consumer as well as a topic. In the meantime, here's a simple addition to the KafkaStateActor test that will get empty lists of consumers.

jisookim0513 commented:
Could you squash the commits?

cvcal commented:
Currently, the call to get the partition offsets with SimpleConsumer is blocking, I'm going to transfer all of those operations to Futures.

jisookim0513 commented:
@cvcal where would you like to have futures in the ActorModel? Not sure if having an execution context in ActorModel is a good idea. Also could you please squash commits into one when you think you won't need to reset to previous commit? It'll be nice to keep the number of commits under 5 per PR, ideally 1 or 2.

cvcal commented:
Ok, I will, I just want to finish this first.

Currently, with a sufficient number of topics and partitions, having to make a simple consumer for each can trigger timeouts, especially with calls like KSGetAllTopics. I thought having futures within the returned value would be the cleanest and most resilient solution. Is there another solution you would prefer? I could also make the KafkaStateActor do some additional work in parallel internally, but if I can't return embedded futures, it would still have to block until everything finished.


> On Aug 13, 2015, at 14:23, Jisoo Kim <notifications@github.com> wrote:
> 
> @cvcal where would you like to have futures in the ActorModel? Not sure if having an execution context in ActorModel is a good idea. Also could you please squash commits into one when you think you won't need to reset to previous commit? It'll be nice to keep the number of commits under 5 per PR, ideally 1 or 2.
> 
> —
> Reply to this email directly or view it on GitHub.
>

jisookim0513 commented:
@cvcal sorry, I have been very busy with other issues so I couldn't get a chance to read through your entire PR. I will try to do it tonight, but for the execution context, you can consider changing KafkaStateActor to LongRunningActor and use longRunningExecutionContext.

cvcal commented:
Ah, I hadn't noticed that LongRunningPoolActor was a trait and could be mixed with BaseQueryCommandActor, thanks!

I'll try to have my update out by tonight then.

jisookim0513 commented:
Cool, thanks. Could you ping me when you have the version you like? I'd like to try it on my laptop as well.

cvcal commented:
@jisookim0513 So this version is better - it still has a problem with creating too many futures because it ends up asking for fresh offsets for a topic every time you ask for the topic description or for the consumer description of a consumer that eats from it - so if a topic or topics gets consumed from a lot, the redundant work piles up and could approach the task-limit. The fix for that is caching the results, but this should work for now.

jisookim0513 commented on app/kafka/manager/KafkaStateActor.scala:
```diff
@@ -218,7 +383,16 @@ class KafkaStateActor(curator: CuratorFramework,
+      case KSGetConsumedTopicDescription(consumer, topic) =>
+        sender ! getConsumedTopicDescription(consumer,topic)
+
+      case KSGetConsumerDescriptions(consumers) =>
```
Where is this used?

cvcal replied:
Which one? KSGetConsumerDescription gets used in the ClusterManagerActor, KSGetConsumers gets used in the KafkaManager, KSGetAllConsumerDescriptions gets used by the BrokerViewCacheActor. I guess KSGetConsumerDescriptions doesn't get used, you're right, I added it automatically, when I was thinking of having a list of valid consumer groups that the user could curate, and forgot to get rid of it.

jisookim0513 commented:
Hmm.. I tried the latest version on your branch, and it worked with my local testing cluster that had less than three topics. However, it doesn't seem to be working with a remote cluster that has quite a bit of topics and partitions. Caching and polling the information sounds like a good idea instead of fetching fresh data every time. However, I am a bit concerned that I wasn't able to see the topic list with the use of futures. Would it be insufficient if we don't have the topic offset information on TopicDescription (and TopicIdentity)? 

@patelh what do you think?

cvcal commented:
It didn't work at all? I'm having a problem with java IO exceptions currently, is the problem you're having linked to ClosedChannelExceptions? I'm trying to fix that right now, so if the problem you were having is unrelated, could you let me know?

jisookim0513 commented:
@cvcal with my local test cluster I remember I was having a problem with ClosedChannelExceptions but I was able to see consumer information and topic offsets. For the remote cluster, the time-out occurred so I couldn't access topic list (only broker list worked).

cvcal commented:
Ok, that's what I thought. 
This should fix it. If you're curious, the problem was that I was creating too many channels (one for each consumer) and was running up against a java limit. I had modeled the offset-getting simpleconsumer code after Kafka's GetOffsetShell, which did not close the connections it created, probably because it could get away with it. In this case, though, the limit was a problem. 
The reason you could still see the brokerlist was that that was the only page that listed information that did not try to get topic or consumer identities, and so it relied only on the channels created for the curator caches, which were still open. Caching delayed the arrival of the channel limit, since it makes fewer simple consumers, and then adding consumer.close() fixed the underlying problem.

cvcal commented:
So I cleaned up the UI a little bit, and updated the README with the new images, here they are for easy access:

First here is the ConsumerListView:
![consumer-list](https://cloud.githubusercontent.com/assets/6541539/9370477/9a6f2afa-469e-11e5-9174-b1bf37a6cef0.png)

The ConsumedTopicView:
![consumed-topic](https://cloud.githubusercontent.com/assets/6541539/9370481/a1aa8c24-469e-11e5-93cd-d845a7c313ba.png)

And the alteration to the TopicView:
<img width="1150" alt="topic" src="https://cloud.githubusercontent.com/assets/6541539/9370487/a7ec74ee-469e-11e5-920f-8cd495aceba9.png">

patelh commented:
Please rebase off master.  Let me know if you need help (please add me as contributor to your repo so I can push to your branch).

patelh commented:
I've started fixing merge conflicts on my local branch.

cvcal commented:
Thanks @patelh! I'm also no longer a contributor to Knewton's repo, but I'll see what I can do.

cvcal commented:
@patelh If knewton were to restore my write access temporarily, could you give me access to where you were fixing the conflicts so that I can push it to this repo?
Thanks for helping out with this!

cvcal commented:
I don't know if this helps,  but if you push your merge conflict fixes to my repository, I can ask Knewton to accept my pull request to their fork. I'm sorry this is more complicated and slow than it could have been. Thanks for your patience, @patelh

patelh commented:
I've pushed a new branch to your repo.  Please push to knewton or close this PR and create a new one or I can create a new PR with your branch (your commits would be preserved).

patelh commented:
@cvcal It may be faster to create a new pull request to merge cvcal branch from your repo to yahoo's.  Please just create a new PR.

cvcal commented:
I agree, I wanted to see if it could happen quickly enough to be worth it, but I doubt it. 
I created https://github.com/yahoo/kafka-manager/pull/131 instead.
